workers/scanner.go

```diff
@@ -0,0 +1,232 @@
+			}
+			if health == 0 {
+				// The file is now fully uploaded and healthy.
+				break
+			}
+			select {
+			case <-ticker.C:
+				s.staticLogger.Tracef("Waiting for '%s' to become fully healthy. Current health: %.2f", skylink, health)
+			case <-deadlineTimer.C:
+				s.staticLogger.Warnf("Skylink '%s' failed to reach full health within the time limit.", skylink)
+				break
+			case <-s.staticTG.StopChan():
+				return
+			}
+		}
+	}
+}
+
+// findAndPinOneUnderpinnedSkylink scans the database for one skylinks which is
+// either locked by the current server or underpinned. If it finds such a
+// skylink, it pins it to the local skyd. The method returns true until it finds
+// no further skylinks to process.
+func (s *Scanner) findAndPinOneUnderpinnedSkylink() (skylink string, sf skymodules.SiaPath, continueScanning bool) {
+	sl, err := s.staticDB.FindAndLockUnderpinned(context.TODO(), s.staticServerName, s.staticMinPinners)
+	if errors.Contains(err, database.ErrSkylinkNotExist) {
+		// No more underpinned skylinks pinnable by this server.
+		return
+	}
+	if err != nil {
+		s.staticLogger.Warn(errors.AddContext(err, "failed to fetch underpinned skylink"))
+		return
+	}
+	defer func() {
+		err = s.staticDB.UnlockSkylink(context.TODO(), sl, s.staticServerName)
+		if err != nil {
+			s.staticLogger.Debug(errors.AddContext(err, "failed to unlock skylink after trying to pin it"))
+		}
+	}()
+	sf, err = s.staticSkydClient.Pin(sl.String())
+	if err != nil {
+		s.staticLogger.Warn(errors.AddContext(err, "failed to pin skylink"))
+		continueScanning = true
+		return
+	}
+	s.staticLogger.Debugf("Successfully pinned %s", sl)
+	return sl.String(), sf, true
+}
+
+// estimateTimeToFull calculates how long we should sleep after pinning the given
+// skylink in order to give the renter time to fully upload it before we pin
+// another one. It returns a ballpark value.
+//
+// This method makes some assumptions for simplicity:
+// * assumes lazy pinning, meaning that none of the fanout is uploaded
+// * all skyfiles are assumed to be large files (base sector + fanout) and the
+//	metadata is assumed to fill up the base sector (to err on the safe side)
+func (s *Scanner) estimateTimeToFull(skylink string) time.Duration {
+	meta, err := s.staticSkydClient.Metadata(skylink)
+	if err != nil {
+		s.staticLogger.Warn(errors.AddContext(err, "failed to get metadata for skylink"))
+		return SleepBetweenPins
+	}
+
+	numChunks := math.Ceil(float64(meta.Length) / chunkSize)
+	// remainingUpload is the amount of data we expect to need to upload until
+	// the skyfile reaches full redundancy.
+	remainingUpload := numChunks*chunkSize*fanoutRedundancy + (baseSectorRedundancy-1)*sectorSize
+	secondsRemaining := remainingUpload / assumedUploadSpeedInBytes
+	return time.Duration(secondsRemaining) * time.Second
+}
+
+// deadline calculates how much we are willing to wait for a skylink to be fully
+// healthy before giving up. It's twice the expected time, as returned by
```

Comment:
`nit` 😁  twice makes it completely arbitrary again, maybe `assumedUploadSpeedInBytes` could be just 10% of our gbps bandwidth and you just escape after the estimated time ran out and set the deadline to the actual estimate